include/deal.II/matrix_free/operators.h

```diff
@@ -328,6 +328,75 @@ namespace MatrixFreeOperators
 
 
 
+  /**
+   * This class implements the operation of the action of a Laplace matrix.
+   *
+   * @author Denis Davydov, 2016
+   */
+  template <int dim, int fe_degree, int n_q_points_1d = fe_degree+1, int n_components = 1, typename Number = double>
+  class LaplaceOperator : public Base<dim, Number>
+  {
+  public:
+
+    /**
+     * Constructor.
+     */
+    LaplaceOperator ();
+
+    /**
+     * The diagonal is approximated by computing a local diagonal matrix per element
+     * and distributing it to the global diagonal. This will lead to a slightly
+     * different diagonal for the case with hanging nodes.
+     */
+    virtual void compute_diagonal ();
+
+    /**
+     * Set heterogeneous scalar coefficient @p scalar_coefficient to be used at
+     * the quadrature points. The Table should be of correct size, consistent
+     * with the total number of quadrature points in <code>dim</code>-dimensions,
+     * controlled by @p n_q_points_1d template parameter.
+     */
+    void set_coefficient(const Table<2, VectorizedArray<Number> > &scalar_coefficient );
+
+    virtual void clear();
+
+  private:
+    /**
+     * Applies the laplace matrix operation on an input vector. It is
+     * assumed that the passed input and output vector are correctly initialized
+     * using initialize_dof_vector().
+     */
+    virtual void apply_add (LinearAlgebra::distributed::Vector<Number>       &dst,
+                            const LinearAlgebra::distributed::Vector<Number> &src) const;
+
+    /**
+     * Applies the Laplace operator on a cell.
+     */
+    template <bool homogeneous>
+    void local_apply_cell (const MatrixFree<dim,Number>                     &data,
+                           LinearAlgebra::distributed::Vector<Number>       &dst,
+                           const LinearAlgebra::distributed::Vector<Number> &src,
+                           const std::pair<unsigned int,unsigned int>  &cell_range) const;
+
+    /**
+     * Apply diagonal part of the Laplace operator on a cell.
+     */
+    template <bool hommogeneous>
+    void local_diagonal_cell (const MatrixFree<dim,Number>                &data,
+                              LinearAlgebra::distributed::Vector<Number>  &dst,
+                              const unsigned int &,
+                              const std::pair<unsigned int,unsigned int>  &cell_range) const;
+
+
+    /**
+     * User-provided heterogeneity coefficient.
+     */
+    SmartPointer<const Table<2, VectorizedArray<Number> >,
+                 LaplaceOperator<dim,fe_degree,n_q_points_1d,n_components,Number> > scalar_coefficient;
```

Comment:
@davydden @bangerth Regarding the coefficient: I have used such constructions in two different applications where the solutions turned out to be different. In one I had exposed the table via a shared pointer (that's where my suggestion came from), in the other one I had only provided a function to pass the coefficient to one index `cell` and one index `q`. This way, the storage is not exposed other than it runs over an outer cell index and an inner q index. This has the advantage that only one class has to hold the coefficient. But it means that you need an additional call to initialize the coefficient field. Furthermore, if one wants to re-use the coefficients in several operators one needs to use the access functions.

In any case, I think that letting a user give the table coefficients is very flexible and covers a wide range of applications. If we ever feel like adding an `std::function` to compute something, that would be straight-forward to add for filling the table. Regarding the table storage I would prefer if the Laplace operator can control the table because having too many objects around is a pain. This is achieved both when only providing an access to a single entry and when doing the shared_ptr construction. I have no strong opinion and probably need to sleep over this for giving a good advice ;-)